### PR TITLE
Enhance catching rules

### PR DESCRIPTION
**Pull request by un-tone**
For https://github.com/StarcounterApps/Website/issues/48.
Related PR: https://github.com/StarcounterApps/Simplified/pull/62.

### CODE REVIEW

**Comment by warpech** [comment]
I am missing one thing. In the new UI where do I set the **catch-all* rule for a surface?

The new feature of enhanced catching rules needs to be applicable for **catch-all** rules too, so that I can say that a particular surface catches all requests that have **mobile** in the `User-Agent` header. Or requests that have **subdomain.mydomain.com** in the `Host` header.

**Comment by un-tone** [comment]
@warpech you can apply headers for catch-all rules in the same way as for other rules: create a rule with an empty URI, add a header for the rule (header key is case-insensitive) and save it.
In the last commit I made some fixes for request handling, please try to check this feature again.

**Comment by un-tone** [comment]
I'm a little confused. Maybe you expected that for catch-all rules there will be some separated form area and elements in UI?

**Comment by warpech** [comment]
> I'm a little confused. Maybe you expected that for catch-all rules there will be some separated form area and elements in UI?

Yes, sorry. I wrongly assumed https://github.com/StarcounterApps/Website/issues/115 is included in  https://github.com/StarcounterApps/Website/pull/135, but it is not.

**Comment by un-tone** [comment]
@PatrykSzwer the conflict is resolved and the PR is re-targeted to `develop`.

File: src/WebsiteProvider/Api/ContentHandlers.cs [hunk 1]
```diff
@@ -104,19 +110,34 @@ protected void RegisterFilter()
                         var wrapper = response.Resource as SurfacePage;
                         var requestUri = request.Uri;
                         var isWrapped = false;
+                        var requestKey = RequestStorage.Put(request);
 
-                        while ((wrapper == null || wrapper.IsFinal == false) && this.HasCatchingRule(requestUri))
+                        try
```
Comment (PatrykSzwer) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: You're using try here but without any exception catching. What about exceptions that can came up here? Shouldn't those be properly handled, not throwing without catching like that and forgot?

File: src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html [hunk 2]
```diff
@@ -1,57 +1,85 @@
...
                         Cancel
                     </button>
-                    <button type="button" class="btn btn-xs btn-default" value="{{model.Create$::click}}"
+                    <button type="button" class="btn btn-xs btn-default" value="{{model.CreateTrigger$::click}}"
                             onmousedown="++this.value;" title="Create catching rule">
                         Create
                     </button>
                 </span>
-                Catching rules
+                Catching rules for {{model.SurfaceName}}
```
Comment (PatrykSzwer) [inline src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html]: I can't see need for telling user again in which surface scope is he right now. I think information on the navigation bar is enough and this is redundant. I would leave just "Catching rules" title.

File: src/WebsiteProvider/Api/ContentHandlers.cs [hunk 3]
```diff
@@ -221,24 +242,45 @@ protected SurfacePage GetSurfacePage(WebTemplate template)
             return page;
```
Comment (PatrykSzwer) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: The method `WrapExternalRequest` is not used anymore any should be removed (lines 205-209, couldn't made comment specifically on them).

File: src/WebsiteProvider/Api/ContentHandlers.cs [hunk 4]
```diff
@@ -221,24 +242,45 @@ protected SurfacePage GetSurfacePage(WebTemplate template)
...
+        private WebUrl FindUrlByHeaders(QueryResultRows<WebUrl> webUrls, Dictionary<string, string> requestHeaders)
+        {
+            if (!webUrls.Any())
+            {
+                return null;
+            }
+            if (webUrls.All(x => !x.Headers.Any()))
+            {
+                return webUrls.First();
+            }
```
Comment (PatrykSzwer) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: I have a question about this part. I don't really get it, in method for finding Url, when there is no any weburl with headers u're returning first one of them? Could You explain to me why this is like that and how should It work?

File: src/WebsiteProvider/Api/ContentHandlers.cs (hunk 1 above)
Comment (un-tone) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Here I need only in `finally` block. Currently I suppose it is an optimal variant to just throw an unhandled exceptions here. If you have any proposals how we should handle these exceptions we could discuss it in separate issue, WDYT?

File: src/WebsiteProvider/Api/ContentHandlers.cs (hunk 3 above)
Comment (un-tone) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Removed, thanks

File: src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html (hunk 2 above)
Comment (un-tone) [inline src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html]: Removed it.

File: src/WebsiteProvider/Api/ContentHandlers.cs (hunk 4 above)
Comment (un-tone) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Yes, it is how it worked before this changes:
`Db.SQL<WebUrl>("SELECT wu FROM Simplified.Ring6.WebUrl wu WHERE wu.Url = ?", url).First`. We just get first appropriate catching rule for current request.
URI of catching rule has a higher priority than headers so if current WebUrl query doesn't have rules with headers then we return the one found only by URI.

**Comment by un-tone** [comment]
I can't see the reason of the TC build failure. @warpech could you take a look at this, please.
It works well on my local machine.

**Comment by warpech** [comment]
TeamCity is getting this error when resolving Nuget:

```
[16:09:34][restore] The 'Newtonsoft.Json 10.0.3' package requires NuGet client version '2.12' or above, but the current NuGet version is '2.8.60717.93'.
[16:09:34][restore] 'NUnit' already has a dependency defined for 'NETStandard.Library'.
[16:09:34][restore] Process exited with code 1
```

I think this is the same problem as https://github.com/Starcounter/level1/issues/3930#issuecomment-314440227

@mmnosek do you know what to do?

**Comment by warpech** [comment]
@un-tone try the same version of Newtonsoft.Json as used here: https://github.com/Starcounter/TPCC/blob/master/TPCC/packages.config#L4

**Comment by warpech** [comment]
Now the error message is just:

```
[18:04:33][restore] 'NUnit' already has a dependency defined for 'NETStandard.Library'.
[18:04:33][restore] Process exited with code 1
```

**Comment by warpech** [comment]
Compare `develop` with `48_enhance-catching-rules`:

screen shot 2017-07-21 at 18 07 53

screen shot 2017-07-21 at 18 07 56

**Comment by un-tone** [comment]
I downgraded back nuget packages for test project. It helped with issue.
Thanks for help @warpech

**Comment by mmnosek** [comment]
> @mmnosek do you know what to do?

Unfortunately, we are pretty much stuck with this problem. Having no option to use different assemblies references makes any kind of "migration project" almost impossible. So we don't have a solution.

**Comment by un-tone** [comment]
@PatrykSzwer thank you! I fixed this bug

**Pull request merged**